classes/reporter.php

```diff
@@ -0,0 +1,747 @@
+                $output .= $OUTPUT->render_from_template('tool_fix_delete_modules/task_report', $reportdata);
+            } else {
+                $output .= $taskreporttitle.PHP_EOL.$reporttables.PHP_EOL;
+            }
+        }
+        return $output;
+    }
+
+    /**
+     * get_diagnosis() - Get the diagnosis of all course_delete_module tasks in either HTML or plain text format.
+     *
+     * @return string
+     */
+    public function get_diagnosis() {
+        global $OUTPUT;
+
+        $output = '';
+        $diagnoses = $this->get_diagnosis_data();
+
+        // Each diagnosis is for a separate adhoc task.
+        foreach ($diagnoses as $diagnosis) {
+            $task = $diagnosis->get_task();
+            // Skip if not in filtered list of taskids.
+            if (!empty($this->querytaskids) && !in_array($task->taskid, $this->querytaskids)) {
+                continue;
+            }
+            $titleword = get_string('diagnosis', 'tool_fix_delete_modules');
+            $diagnosistitle = $this->get_word_task_module_string($titleword, $task);
+            if ($this->ishtmloutput) {
+                $symptoms = $diagnosis->get_symptoms();
+                $diagnosistable = $this->get_htmltable($symptoms, [get_string('symptoms', 'tool_fix_delete_modules')]);
+                $diagnosisbutton = $this->get_fix_button($diagnosis);
+                $diagnosisdata = ['title' => $diagnosistitle, 'table' => $diagnosistable, 'button' => $diagnosisbutton];
+                $output .= $OUTPUT->render_from_template('tool_fix_delete_modules/task_diagnosis', $diagnosisdata);
+            } else {
+                $symptoms = $diagnosis->get_symptoms();
+                if (empty($symptoms)) { // If no symptoms, then add a "It's all Good" string for CLI output.
+                    $symptoms = array(get_string(diagnosis::GOOD, 'tool_fix_delete_modules'));
+                    $heading = array(); // No heading needed.
+                } else {
+                    $heading = [get_string('symptoms', 'tool_fix_delete_modules')];
+                }
+                $diagnosistable = $this->get_texttable($symptoms, $heading);
+                $output .= $diagnosistitle.PHP_EOL.$diagnosistable;
+            }
+        }
+        return $output;
+    }
+
+    /**
+     * get_diagnosis_data() - Helper function for get_diagnosis(); gets the diagnosis object returned from diagnoser.
+     *
+     * @param array $taskids - optional array of ints (task id(s) to be fixed); empty array means no filter.
+     *
+     * @return array of diagnosis
+     */
+    private function get_diagnosis_data(array $taskids = null) {
+        $diagnoses = array();
+        $deletetaskslist = new delete_task_list($this->minimumfaildelay);
+        $deletetasks = $deletetaskslist->get_deletetasks();
+        foreach ($deletetasks as $taskid => $deletetask) {
+            if (empty($taskids) || in_array($taskid, $taskids)) { // If no filter or is in the filter.
```

Comment:
Is this method (and `make_fix()` ) being called anywhere with no filters/tasks set? If not I would just stick with `$taskids` as a required param and do nothing if it's not passed (or throw an exception).

Reply:
Yes, it is, in `get_diagnosis()`

Reply:
Could you please provide the link?

Reply:
https://github.com/catalyst/moodle-tool_fix_delete_modules/blob/dev-objectoriented-35/classes/reporter.php#L111